Fix tests for nomad/client/driver/executor package to work on Windows.

ChrisHines opened the pull request:
With these changes, all tests for this package pass on Windows.

ChrisHines commented:
It looks to me like the test failures were caused by an unrelated part of the code, something to do with the client/fingerprint package.

dadgar commented:
Shouldn't this line also use the TestMain: https://github.com/hashicorp/nomad/pull/462/files#diff-4ec3b758ae80396d0810cd04db40e05eL235

dadgar commented on client/driver/executor/test_harness_test.go:
```diff
@@ -14,6 +15,57 @@ import (
 )
 
+func TestMain(m *testing.M) {
+	switch os.Getenv("TEST_MAIN") {
```
Please add a comment that we switch the behavior depending on the environment variable and why we do this.

dadgar commented on client/driver/executor/test_harness_test.go:
```diff
@@ -14,6 +15,57 @@ import (
+		os.Exit(1)
+	}
+	args := os.Args[1:]
+	shift := func() string {
```
Comment on what shift does

dadgar commented on client/driver/executor/test_harness_test.go:
```diff
@@ -14,6 +15,57 @@ import (
+	for len(args) > 0 {
+		switch cmd := shift(); cmd {
+		case "sleep":
+			if len(args) < 1 {
```
I would prefer if you put a comment on each case describing the behavior and how many arguments they take etc. It is essentially a function and we should document it as such.

dadgar commented on client/driver/executor/test_harness_test.go:
```diff
@@ -45,9 +97,10 @@ func testExecutor(t *testing.T, buildExecutor func() Executor, compatible func(*
-		return b
+		e := buildExecutor()
+		SetCommand(e, name, args)
+		e.Command().Env = append(os.Environ(), "TEST_MAIN=app")
```
Worth sticking `TEST_MAIN=app` into a const and documenting it

dadgar commented on client/driver/executor/test_harness_test.go:
```diff
@@ -79,7 +132,7 @@ func Executor_Start_Invalid(t *testing.T, command buildExecCommand) {
 
 func Executor_Start_Wait_Failure_Code(t *testing.T, command buildExecCommand) {
-	e := command("/bin/date", "-invalid")
+	e := command(os.Args[0], "fail")
```
What do you think about putting `os.Args[0]` into a a variable. It could then be documented and be given a more useful name to make it clear that we are calling ourselves.

dadgar commented on client/driver/executor/test_harness_test.go:
```diff
@@ -14,6 +15,57 @@ import (
+	}
+}
+
+func appMain() {
```
Can we move this somewhere so that it can be shared between executor and spawn

ChrisHines replied:
We should be able to move almost all of the TestMain/appMain code into a helper package. We will still need a minimal TestMain in each package that needs appMain, but it can get all of its functionality by calling another function like so:

```go
func TestMain(m *testing.M) {
    testhelp.Run(m)
}
```

Where `testhelp.Run` contains the `switch os.Getenv ...` code.

ChrisHines commented:
Thanks for the review comments above. I agree with everything and will update this PR when I get a chance.

There is a very good reason why `Executor_Open_Invalid` does not call back into the test binary. That test is validating the behavior when the user program doesn't exist at all. I will add a comment explaining the reason for the exception.

ChrisHines commented:
OK. I have rebased and added a commit based on the above review comments.

I have not yet factored the appMain code into a separate package. Would you prefer I do that and update this PR and #442 before we merge? It seems easier to factor out the common code later to avoid a dependency between the PRs, but I am happy to do it either way.

dadgar commented:
> There is a very good reason why Executor_Open_Invalid does not call back into the test binary. That test is validating the behavior when the user program doesn't exist at all. I will add a comment explaining the reason for the exception.

This is actually not the point of that test. If you look here: https://github.com/hashicorp/nomad/pull/462/files?diff=split#diff-4ec3b758ae80396d0810cd04db40e05eR337 you can see that the alloc directory is destroyed which removes the task state information. So what the test is really checking is that an open to an invalid directory will return an error.

dadgar commented:
As for the shared code between the two PRs, I am okay with merging these in and having a follow up PR. Will wait for an update from you on that one test and then merge!

dadgar commented:
Windows/OS X seem to work but linux is failing: http://pastebin.com/PeeGwD20
This was on the Vagrantbox in the project.

ChrisHines commented:
>This is actually not the point of that test.

Oops, I was thinking of `Executor_Start_Invalid`. Yes, I agree that `Executor_Open_Invalid` should use `TestMain`. I will fix that when I get a chance.

As for failing on Linux, I will have to reproduce and investigate.

ChrisHines commented:
I was wondering why `Executor_Open_Invalid` wasn't failing on Windows, since it wasn't using the portable `appMain` code. Reason: It wasn't being called from `testExecutor`!

Fixing that oversight as well as using `appMain`, I find that the test now fails on Windows. I suspect that `alloc.Destroy` may not actually be working as expected, but need to investigate more.

ChrisHines commented:
All of the tests pass for me on Windows as well as Ubuntu 15.10 (inside VitualBox). The TravisCI test passes for the client/driver/executor package, but fails elsewhere.

ChrisHines commented:
@dadgar please take another look.

dadgar commented:
Closing in favor of #497
